python/pipeline/pupil.py

```diff
@@ -1151,6 +1153,352 @@ def plot_fitting(key, start, end=-1, fit_type='Circle', fig=None, ax=None, mask_
+            leftlid_x, leftlid_y = (FittedPupil.EyePoints & key & 'label="eyelid_left"').fetch1('x','y')
+            rightlid_x, rightlid_y = (FittedPupil.EyePoints & key & 'label="eyelid_right"').fetch1('x','y')
+            eyelid_widths = [np.sqrt((lx-rx)**2 + (ly-ry)**2) for lx,ly,rx,ry in 
+                             zip(leftlid_x, leftlid_y, rightlid_x, rightlid_y)]
+            
+            ## Assume that the true eyelid width is the median
+            eyelid_width_in_px = np.nanmedian(eyelid_widths)
+            
+            ## Define mm/px
+            conversion_factor = eyelid_width_in_mm / eyelid_width_in_px
+            
+            ## Sanity check extrapolated size
+            pupil_source = (PupilFilterMethod & key).fetch('pupil_source')
+            
+            if pupil_source == 'Circle':
+                source_table = FittedPupil.Circle
+            elif pupil_source == 'Ellipse':
+                source_table = FittedPupil.Ellipse
+            
+            radius_type = (PupilFilterMethod & key).fetch1('radius_type')
+            pupil_radii = (source_table & key).fetch(radius_type)
+            max_pupil_diameter_px = 2*np.nanpercentile(pupil_radii,99.9)
+            
+            if max_pupil_diameter_px > eyelid_width_in_px:
+                pupil_diameter_mm = round(max_pupil_diameter_px * conversion_factor, 3)
+                msg = f"Pupil diameter {pupil_diameter_mm}mm exceeds assumed eyelid width {eyelid_width_in_mm}mm."
+                raise PipelineException(msg)
+            
+        else:
+            
+            msg = f"Unit conversion method id {key['period_method_id']} is not supported."
+            raise PipelineException(msg)
+            
+        self.insert1({**key, 'conversion_factor': conversion_factor})
+
+
+@schema
+class PupilPeriodMethod(dj.Lookup):
+    definition = """ # Method to extract periods of dilation/constriction in pupil trace
+    period_method_id              : tinyint unsigned    # index of pupil period method
+    ---
+    period_method_description     : varchar(256)
+    """
+    
+    contents = [[1, 'Peak detection followed by filtering out events < 1sec or with an average speed < 0.02mm/sec. Any event comprised of over 15% NaNs is automatically dropped.'],
+                ]
+    
+
+@schema
+class PupilPeriods(dj.Computed):
+    definition = """ # Scan level statistics on dilations & constrictions
+    -> ProcessedPupil
+    -> PupilPeriodMethod
+    -> PupilUnitConversion
+    ---
+    dilation_freq                 : float               # dilations per second (within valid start/end)
+    constriction_freq             : float               # constrictions per second (within valid start/end)
+    num_pupil_periods             : smallint unsigned   # total number of dilation and constriction periods
+    """
+
+    class DilationConstriction(dj.Part):
+        definition = """ # Single period of dilation or constriction
```

Comment:
does the period_index specify whether it's dilation or constriction? if so, the comment should say that. If not, how do you tell the difference between a constriction and dilation period?